llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -234,76 +209,250 @@ def __init__(
+
+        # Validate the interval (how often to launch eval runs)
+        self.interval = validate_interval(interval,
+                                          self.training_params['save_interval'])
+
+        # Configures how often to check for new checkpoints
+        self.check_interval = Time(max(self.interval.value // 5, 1),
+                                   self.interval.unit)
+
+        # Keep track of checkpoints by interval that have already been evaled
+        # Format: {interval: (checkpoint, run_name)}
+        self.checkpoints_evaled: Dict[Time, Tuple[str, str]] = {}
+
+        # Scheduling is based on the check interval, while _get_checkpoints_and_launch_runs
+        # will only launch runs at the interval
+        self.is_at_check_interval = create_interval_scheduler(
+            self.check_interval,
+            # There is a custom close to ensure that the final checkpoint
+            # (which is the most important) is evaled after it is written
+            include_end_of_training=False,
         )
 
+        log.info('Initialized AsyncEval callback. Will generate runs at ' +
+                 f'interval {interval}, checking at {self.check_interval}')
+
+    def state_dict(self) -> Dict[str, Any]:
+        checkpoints_evaled = []
+        for i, (c, rn) in self.checkpoints_evaled.items():
+            interval_dict = {
+                'value': i.value,
+                'unit': i.unit.value,
+            }
+            checkpoints_evaled.append((interval_dict, c, rn))
+
+        return {
+            'checkpoints_evaled': checkpoints_evaled,
+        }
+
+    def load_state_dict(self, state_dict: Dict[str, Any]):
+        previous_checkpoints_evaled = state_dict.get('checkpoints_evaled', [])
+        if previous_checkpoints_evaled:
+            for (i, c, rn) in previous_checkpoints_evaled:
+                interval = Time(i['value'], TimeUnit(i['unit']))
+                self.checkpoints_evaled[interval] = (c, rn)
+
+            log.info(
+                f'Loaded previous checkpoints evaled: {self.checkpoints_evaled}'
+            )
+
+    @staticmethod
+    def _get_ready_sharded_checkpoints(
+        checkpointer_checkpoints: Dict[str, Timestamp],
+        remote_files: List[str],
+    ) -> Dict[str, Timestamp]:
+        """Identify checkpoints ready to be evaled based on remote files.
+
+        This has special logic for sharded checkpoints to consider checkpoints composed
+        of multiple shards (one per gpu) and metadata
+
+        Args:
+            checkpointer_checkpoints: All checkpoints from the checkpointer state
+            remote_files: List of remote files in the save folder
+
+        Returns:
+            Dict of checkpoints that are complete and ready to be evaled
+        """
+        # Count the number of shards for each checkpoint group
+        remote_file_group_counts = Counter()
+        for f in remote_files:
+            interval_path = Path(f).parts[-2]
```

Comment:
I think this is not totally guaranteed, since we let users customize the paths. Should we instead be taking taking the full path without the filename? or does that break your logic elsewhere?